### PR TITLE
fix(rome_js_formatter): change parenthesized expression formatting

### LINKED ISSUE #2427: Parenthesized binary expression
Prettier doesn't add an additional indent level if a binary expression is wrapped by a parenthesized expression but Rome does. [Playground](https://play.rome.tools/?lineWidth=80&indentStyle=space&quoteStyle=double&indentWidth=2&typescript=true&jsx=false&sourceType=module#KAAKACAAIABhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhACAAJgAmAAoAIAAgAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAIAAmACYACgAgACAAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwAgACYAJgAKACAAIABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkACAAJgAmAAoAIAAgAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUACgApACgAKQA7AAoACgAoAAoAIAAgAGEAYQAgACYAJgAKACAAIABiAGIAIAAmACYACgAgACAAYwBjACAAJgAmAAoAIAAgAGQAZAAgACYAJgAKACAAIABlAGUACgApACgAKQA7AAoACgAvAC8AIABJAG4AZABlAG4AdAAnAHMAIABhAHMAIAB1AHMAdQBhAGwAIABpAGYAIABpAHQAJwBzACAAYQAgAHAAYQByAGUAbgB0AGgAZQBzAGkAegBlAGQAIABzAGUAcQB1AGUAbgBjAGUAIABlAHgAcAByAGUAcwBzAGkAbwBuAAoAKAAKACAAIABhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhACAAKwAKACAAIABiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiACAAKwAKACAAIABjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjACAAKwAKACAAIABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkACAAKwAKACAAIABlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlACwAIAAiAHQAZQBzAHQAIgAKACkAKAApADsACgAKACgACgAgACAAYQBhACAAKwAKACAAIABiAGIAIAArAAoAIAAgAGMAYwAgACsACgAgACAAZABkACAAKwAKACAAIABlAGUACgApACgAKQA7AAoACgAoAAoAIAAgAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAIAAmACYACgAgACAAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgAgACYAJgAKACAAIABjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjACAAJgAmAAoAIAAgAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAIAAmACYACgAgACAAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQAKACkAKAApACgAKQAoACkAOwAKAAoALwAvACAARABvAGUAcwBuACcAdAAgAGEAcABwAGwAeQAgAGYAbwByACAAYwBhAGwAbAAgAGEAcgBnAHUAbQBlAG4AdABzAAoAYwBhAGwAbAAoAAoAIAAgAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAIAAmACYACgAgACAAIAAgAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAIAAmACYACgAgACAAIAAgAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAIAAmACYACgAgACAAIAAgAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAIAAmACYACgAgACAAIAAgAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUACgApACgACgAgACAAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQAgACYAJgAKACAAIAAgACAAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgAgACYAJgAKACAAIAAgACAAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwAgACYAJgAKACAAIAAgACAAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZAAgACYAJgAKACAAIAAgACAAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQAKACkAKAAKACAAIABhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhAGEAYQBhACAAJgAmAAoAIAAgACAAIABiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiAGIAYgBiACAAJgAmAAoAIAAgACAAIABjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjAGMAYwBjACAAJgAmAAoAIAAgACAAIABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkAGQAZABkACAAJgAmAAoAIAAgACAAIABlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAGUAZQBlAAoAKQA7AAoA)

## Input

```ts
(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
  bbbbbbbbbbbbbbbbbbbbbbbbb &&
  ccccccccccccccccccccccccc &&
  ddddddddddddddddddddddddd &&
  eeeeeeeeeeeeeeeeeeeeeeeee
)();

(
  aa &&
  bb &&
  cc &&
  dd &&
  ee
)();

// Indent's as usual if it's a parenthesized sequence expression
(
  aaaaaaaaaaaaaaaaaaaaaaaaa +
  bbbbbbbbbbbbbbbbbbbbbbbbb +
  ccccccccccccccccccccccccc +
  ddddddddddddddddddddddddd +
  eeeeeeeeeeeeeeeeeeeeeeeee, "test"
)();

(
  aa +
  bb +
  cc +
  dd +
  ee
)();

(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
  bbbbbbbbbbbbbbbbbbbbbbbbb &&
  ccccccccccccccccccccccccc &&
  ddddddddddddddddddddddddd &&
  eeeeeeeeeeeeeeeeeeeeeeeee
)()()();

// Doesn't apply for call arguments
call(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
);

```

## Prettier

```ts
(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
  bbbbbbbbbbbbbbbbbbbbbbbbb &&
  ccccccccccccccccccccccccc &&
  ddddddddddddddddddddddddd &&
  eeeeeeeeeeeeeeeeeeeeeeeee
)();

(aa && bb && cc && dd && ee)();

// Indent's as usual if it's a parenthesized sequence expression
(aaaaaaaaaaaaaaaaaaaaaaaaa +
  bbbbbbbbbbbbbbbbbbbbbbbbb +
  ccccccccccccccccccccccccc +
  ddddddddddddddddddddddddd +
  eeeeeeeeeeeeeeeeeeeeeeeee,
"test")();

(aa + bb + cc + dd + ee)();

(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
  bbbbbbbbbbbbbbbbbbbbbbbbb &&
  ccccccccccccccccccccccccc &&
  ddddddddddddddddddddddddd &&
  eeeeeeeeeeeeeeeeeeeeeeeee
)()()();

// Doesn't apply for call arguments
call(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
);

```

## Rome 

```ts
(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)();

(aa && bb && cc && dd && ee)();

// Indent's as usual if it's a parenthesized sequence expression
(
  aaaaaaaaaaaaaaaaaaaaaaaaa +
    bbbbbbbbbbbbbbbbbbbbbbbbb +
    ccccccccccccccccccccccccc +
    ddddddddddddddddddddddddd +
    eeeeeeeeeeeeeeeeeeeeeeeee, "test"
)();

(aa + bb + cc + dd + ee)();

(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee
)()()();

// Doesn't apply for call arguments
call(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee,
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee,
)(
  aaaaaaaaaaaaaaaaaaaaaaaaa &&
    bbbbbbbbbbbbbbbbbbbbbbbbb &&
    ccccccccccccccccccccccccc &&
    ddddddddddddddddddddddddd &&
    eeeeeeeeeeeeeeeeeeeeeeeee,
);
```

## Expected

Rome's formatting to match Prettier's

### PR DESCRIPTION
**Pull request by yassere**
## Summary

This PR increases Prettier alignment by:
- Preventing binary expressions and sequence expressions from adding indentation when parenthesized
- Preventing sequence expressions from being soft block indented when parenthesized except for when they're a return argument or a unary operand.

Closes #2427

TODO: This introduces a `check_reformat` instability for `binaryish_expression` and `logical_expression` that I still need to troubleshoot.

## Test Plan

Confirm snapshots move closer to Prettier output.

### CODE REVIEW

File: crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap [hunk 1]
```diff
@@ -74,15 +72,13 @@ function a() {
...
-		____________third
-	),
+	something: (____________first,
+	____________second,
+	____________third,
+	____________third,
+	____________third,
+	____________third,
+	____________third),
 };
```
Comment (yassere) [inline crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:82]: This isn’t really an improvement, but trying to fix it in this PR could conflict with #2627
 Our current formatting for this already doesn’t match Prettier’s though.

File: crates/rome_js_formatter/tests/specs/prettier/js/template-literals/sequence-expressions.js.snap [hunk 2]
```diff
@@ -10,9 +11,8 @@ expression: sequence-expressions.js
 
 # Output
 ```js
-`111111111 222222222 333333333 444444444 555555555 666666666 777777777 ${(
-  1, 2
-)}`;
+`111111111 222222222 333333333 444444444 555555555 666666666 777777777 ${(1,
+2)}`;
```
Comment (yassere) [inline crates/rome_js_formatter/tests/specs/prettier/js/template-literals/sequence-expressions.js.snap:15]: This isn’t an improvement, but this might be better handled as part of separate PR that brings template literal formatting into closer alignment with Prettier.

File: crates/rome_js_formatter/src/js/expressions/parenthesized_expression.rs [hunk 3]
```diff
@@ -174,3 +154,26 @@ fn not_binaryish_expression(node: &JsSyntaxNode) -> bool {
...
+            matches!(
+                parent_kind,
+                Some(JsSyntaxKind::JS_RETURN_STATEMENT)
+                    | Some(JsSyntaxKind::JS_UNARY_EXPRESSION)
+                    | Some(JsSyntaxKind::JS_ARROW_FUNCTION_EXPRESSION)
+            )
+        }
+        _ => true,
+    };
+    Ok(requires_indent)
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/expressions/parenthesized_expression.rs:178]: Nit: It should be possible to match directly on the node rather than going over syntax


```
match node.expression() {
  JsAnyExpression::JsStringLiteralExpression(_) => false,
  ...
}
```

File: crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap [hunk 4]
```diff
@@ -51,15 +51,13 @@ const f = () => (
...
-	____________first,
-		____________second,
-		____________third,
-		____________third,
-		____________third,
-		____________third,
-		____________third
-);
+(____________first,
+____________second,
```
Comment (MichaReiser) [inline crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:55]: Wasn't this case handled by the code that you have now removed (and mentioned ema?)

**Review by MichaReiser — COMMENTED** [review]
Nice!

File: crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap (hunk 4 above)
Comment (yassere) [reply crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:55]: In this case, Prettier would just omit the parentheses. But if you force the parentheses to remain like in your original example, Prettier doesn't block indent the sequence expression. The indentation in your example (copied below) came from the binary expression.
```
(aaaaaaaaaaaaaaaaaaaaaaaaa +
  bbbbbbbbbbbbbbbbbbbbbbbbb +
  ccccccccccccccccccccccccc +
  ddddddddddddddddddddddddd +
  eeeeeeeeeeeeeeeeeeeeeeeee,
"test")();
```

Another demonstration is when you assign a sequence expression. I'll add a test case for that. 

Before this PR:
```js
foo =
	(
		____________first,
			____________second,
			____________third,
			____________third,
			____________third,
			____________third,
			____________third
	);
```

After this PR and also Prettier:
```js
foo =
	(____________first,
	____________second,
	____________third,
	____________third,
	____________third,
	____________third,
	____________third);
```

**Comment by yassere** [comment]
> Nice!

It's not very nice until I resolve the reformat instability. 😅
But I may look into how we can implement centralized parentheses insertion rather than just handling it in the binaryish expression formatting rules, because that's a much better solution long-term.

File: crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap (hunk 4 above)
Comment (ematipico) [reply crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:55]: Yeah, I don't know. Given what we have and what prettier does, do we really need to match it? Doesn't seem that good honestly

File: crates/rome_js_formatter/src/js/expressions/sequence_expression.rs [hunk 5]
```diff
@@ -16,25 +16,10 @@ impl FormatNodeFields<JsSequenceExpression> for FormatNodeRule<JsSequenceExpress
...
-                            | JsSyntaxKind::JS_PROPERTY_OBJECT_MEMBER
-                    )
-                )
-            } else {
-                false
-            }
+            matches!(
+                parent.kind(),
+                JsSyntaxKind::JS_RETURN_STATEMENT | JsSyntaxKind::JS_PARENTHESIZED_EXPRESSION
+            )
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_formatter/src/js/expressions/sequence_expression.rs:22]: [Cases like this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEq+AFAJT7D4BOcM6VUJAOlAIbsedfecA0LAIyHCRosSP5Yw0mbLnyZkgCYrVa9RrWS4O3Xv0G9LUgF8QvEBAAOuaAGdkoVlSoQA7gAVnCBylYAbN1YATwcLASpWMABrWgBlVgBbOAAZHCg4ZFQAuzhwyJj4qyj0gHNkGCp0PJA4RIE4FUaU1ihS9FZSuAAxCCpE1hhcNuQQVnQYCHMQAAsYRP8AdRmceDtisDg4n1WcADdV4NGwOzCQdNyqGA9I0oGsnJqAKzsADziy-zgARXQIeAe-lyFmKVEuowErAa-mmVio6RgixwShgM2QAA4AAwg1y5RaRKyjOFwS57TIWACOf3gN2svjGdgAtBlGo1pjQqTgaDdOvckNkgTVcokcBUqkLPj9qZl+Y8LDAoUiUWikAAmeWRHD+MoAYQgiT5tTsAFZpuhcgAVKG+AXAkB7aoASSgSgQMDiYHhNgAgi64jBgl9AbkTCYgA)

File: crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap (hunk 4 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:55]: I"m always in favor of deleting code :D I don't think it's that terrible, nice and compact. But to phrase this differently. I don't think it's that bad that it's reason enough for us to diverge, especially if it means we can delete code.

Comment (yassere) [reply crates/rome_js_formatter/tests/specs/js/module/expression/sequence_expression.js.snap:55]: I expect that this sort of code is uncommon enough in the wild that we don't need to be terribly concerned with how it looks, so I'm also in favor of just simplifying our rules here and matching Prettier. If we ever discover enough of an impact on real-world projects, we can reconsider in the future.
